components/scream/src/physics/cld_fraction/atmosphere_cldfraction.cpp

```diff
@@ -0,0 +1,157 @@
+  // Nothing to do here
+}
+
+// =========================================================================================
+void CldFraction::set_grids(const std::shared_ptr<const GridsManager> grids_manager)
+{
+  using namespace ekat::units;
+
+  // The units of mixing ratio Q are technically non-dimensional.
+  // Nevertheless, for output reasons, we like to see 'kg/kg'.
+  auto Q = kg/kg;
+  Q.set_string("kg/kg");
+  Units nondim(0,0,0,0,0,0,0);
+
+  const auto& grid_name = m_cldfraction_params.get<std::string>("Grid");
+  auto grid  = grids_manager->get_grid(grid_name);
+  m_num_cols = grid->get_num_local_dofs(); // Number of columns on this rank
+  m_num_levs = grid->get_num_vertical_levels();  // Number of levels per column
+  m_nk_pack  = ekat::npack<Spack>(m_num_levs);
+
+  // Define the different field layouts that will be used for this process
+  using namespace ShortFieldTagsNames;
+
+  // Layout for 3D (2d horiz X 1d vertical) variable defined at mid-level and interfaces 
+  FieldLayout scalar3d_layout_mid { {COL,LEV}, {m_num_cols,m_num_levs} };
+
+  // Set of fields used strictly as input
+  m_required_fields.emplace("qi",   scalar3d_layout_mid, Q,      grid_name);
+  m_required_fields.emplace("alst", scalar3d_layout_mid, nondim, grid_name);
+
+  // Set of fields used strictly as output
+  m_computed_fields.emplace("ast",   scalar3d_layout_mid, nondim, grid_name);
+  m_computed_fields.emplace("aist",  scalar3d_layout_mid, nondim, grid_name);
+
+  // Set of fields used as input and output
+  // - There are no fields used as both input and output.
+
+}
+
+// =========================================================================================
+void CldFraction::initialize_impl (const util::TimeStamp& t0)
+{
+  m_current_ts = t0;
+}
+
+// =========================================================================================
+void CldFraction::run_impl (const Real dt)
+{
+/* Luca, do we need any of the following code?
+  // Copy inputs to host. Copy also outputs, cause we might "update" them, rather than overwrite them.
+  for (auto& it : m_cldfraction_fields_in) {
+    it.second.sync_to_host();
+  }
+  for (auto& it : m_cldfraction_fields_out) {
+    it.second.sync_to_host();
+  }
+
+  // Copy outputs back to device
+  // LB: why?!?
+  for (auto& it : m_cldfraction_fields_out) {
+    it.second.sync_to_dev();
+  }
+*/
+
+  
+  // Calculate ice cloud fraction and total cloud fraction given the liquid cloud fraction
+  // and the ice mass mixing ratio. 
+  auto qi   = m_cldfraction_fields_in["qi"].get_reshaped_view<const Pack**>();
+  auto alst = m_cldfraction_fields_in["alst"].get_reshaped_view<const Pack**>();
+  auto aist = m_cldfraction_fields_out["aist"].get_reshaped_view<Pack**>();
+  auto ast  = m_cldfraction_fields_out["ast"].get_reshaped_view<Pack**>();
+
+  cldfracF::cldfraction_main(m_num_cols,m_num_levs,qi,alst,aist,ast);
+
+  // Get a copy of the current timestamp (at the beginning of the step) and
+  // advance it,
```

Comment:
Why does the cldfraction process care about the timestep?

Reply:
I guess the only question is about updating the field time stamp.  Since cld_fraction will be the only process to update cld_frac_i and cld_frac_tot we need to have it grab the timestamp and use it in line 95.  @bartgol , am I correct here?

Reply:
It seems like if all FM variables need a timestamp, it would be easier to just pass the timestep they should use into all processes and have the timestep update occur at the top level rather than in individual processes??? I feel like I'm missing something here.

Reply:
Well, we are not yet _really_ using the time stamp, but _ideally_ one should use the timestamp to figure out when a variable was last updated (at least on rank, we are not syncing time stamps across MPI, maybe we should, I don't know).

If we update all time stamps from the top level, the functionality kind of loses value. If each atm proc updates the ts of only what got updated during that call, maybe it can be more useful. That is, assuming that _some_ variables are not updated during _some_ calls. I'm thinking of something like "ah, it was night, so rad didn't update blah this time step". Maybe this never happens though, and everything is always updated. I don't know. If that's the case (i.e., everything always updated), then I think we could (should?) remove the time stamp from the fields.

Put it another way: what use can you envision from having the ability of checking the time stamp of a field?

Reply:
> what use can you envision from having the ability of checking the time stamp of a field?

I'm not sure why we're doing this. I think the idea was that we could check for race conditions when procs were parallel split? Like "don't update the state until all the expected processes are updated"? Otherwise I'm not sure when we'd use this. I don't think we will check that SW radiation isn't called at night, for example. Overall, I'm unsure this timestamp thing will be that helpful. Maybe I'm just suffering from a case of the grumpies.

> If we update all time stamps from the top level, the functionality kind of loses value.

I'm not sure I agree. As I see it, we can pass all processes the current timestep and if they update a variable, they assign that new timestamp to it. If not, they don't.

Reply:
> I'm not sure I agree. As I see it, we can pass all processes the current timestep and if they update a variable, they assign that new timestamp to it. If not, they don't.

I agree. I guess I misunderstood what you meant with "have the timestep update occur at the top level rather than in individual processes".  I thought you meant by the AD, or something. If the individual procs update the ts of what they update, then it does have some meaning.

Reply:
> I'm not sure why we're doing this. I think the idea was that we could check for race conditions when procs were parallel split? Like "don't update the state until all the expected processes are updated"? Otherwise I'm not sure when we'd use this. I don't think we will check that SW radiation isn't called at night, for example. Overall, I'm unsure this timestamp thing will be that helpful. Maybe I'm just suffering from a case of the grumpies.

Yeah, I'm not sure either. It _might_ help if we had 'related' vars (like T and Theta), and you want to check that the one you want to use is "more recent" than the other (otherwise you might be using stale data). I guess I assumed some vars might go through a few timesteps without being "actually" updated (e.g., solve rad only every other atm step, or something). But I don't think that will happen.

For now, we can keep the ts, I guess, but we should think about whether it can help or not. If not, we should purge it.